### PR TITLE
[READY] Plumbing: Endgame

### PR DESCRIPTION
**Pull request by Time-Green**
:cl:
add: Plumbing equipment is now available through the medical protolathe. It's under medical machinery boards
add: Add's a new chemistry area on metastation for chemical factories. It's located left of main surgery. 
tweak: Chemistry has been turned into the Apothecary. It's basically normal chemistry, but Medical Doctors have acces aswell.
tweak: The smoke machine is now plumbing compatible. Put this knowledge to good use.
sprite: Thanks to @CRITAWAKETS for the plumbing RCD sprite!
/:cl:

Finally adds plumbing to the game!

Chemistry has been turned into the apothecary. For all intents and purposes it's old chem, but doctors have acces.

West of main surgery is now the new chemistry factory area. 

I've also had to move stuff around alot, because you can't just add a big new room without having stuff in the way.

Currently only on metastation. I'm completely drained from adding it to meta alone and doing all maps right now will burn me out.

[Wiki Article](https://tgstation13.org/wiki/Guide_to_plumbing)

image

### CODE REVIEW

**Comment by ShizCalev** [comment]
`Chemistry has been turned into the apothecary. For all intents and purposes it's old chem, but doctors have acces.`

atomize??

**Comment by Time-Green** [comment]
> `Chemistry has been turned into the apothecary. For all intents and purposes it's old chem, but doctors have acces.`
> 
> atomize??

What do you mean?

**Comment by Mey-Ha-Zah** [comment]
Thats... actually quite brilliant. Give the chemists access to equipment to make industrial quantities of meth, and leave doctors to the basic slave machines.

**Comment by Shadowflame909** [comment]
What did it cost?

Time-Greens sanity

**Comment by IndieanaJones** [comment]
Nice to see some more crew content added, always a plus.

File: code/__DEFINES/access.dm
Comment (Angustmeta) [inline code/__DEFINES/access.dm]:
```suggestion
#define ACCESS_APOTHECARY 59
```
x10

Comment (Angustmeta) [inline code/__DEFINES/access.dm]:
```suggestion
#define ACCESS_APOTHECARY 59
```
In several different places

**Comment by Tharcoonvagh** [comment]
The heck is plumbing?

**Comment by 81Denton** [comment]
https://github.com/tgstation/tgstation/pulls?q=is%3Apr+is%3Amerged+plumbing

**Comment by Tharcoonvagh** [comment]
nice

**Comment by CRITAWAKETS** [comment]
I would use this a lot for mechandise. After all, anything can be bought and sold for the right O I L

**Comment by Yttriums** [comment]
Can we call the common chem room Pharmaceuticals or something? I don't feel like a space station would use the name apothecary.

**Comment by Time-Green** [comment]
WIP because I made a few map fucko wuckos

**Comment by Iatots** [comment]
Runtime station already has an empty testroom, you need only add the parts in a corner really.

File: code/modules/research/techweb/all_nodes.dm
Comment (CRITAWAKETS) [inline code/modules/research/techweb/all_nodes.dm]: Typo. Its sewage control.

Comment (CRITAWAKETS) [inline code/modules/research/techweb/all_nodes.dm]: Typo. Its sewage control instead of sewage controle.

**Review by CRITAWAKETS — CHANGES_REQUESTED** [review]
I hate typos. I found one in the tech node description.

File: code/modules/research/techweb/all_nodes.dm
Comment (Time-Green) [reply code/modules/research/techweb/all_nodes.dm]: * It's

Comment (Time-Green) [reply code/modules/research/techweb/all_nodes.dm]: I hate typos. It's "it's" instead of "its".

**Comment by Fhaxaris** [comment]
❌Bombing medical
✔Bombing chem factory

This looks very exciting.

**Comment by Nabski89** [comment]
Why does this entire thing need to be an impenetrable windowless R-Wall fortress?

Add some more windows so you don't get the virology problem where someone works alone in a secure room for 20 minutes only to emerge with with the ultimate death scheme.

**Comment by Time-Green** [comment]
I can’t add more windows, they’ll simply lead to maint. 

I went with r-walls, because it would make sense to contain the explosions. It doesn’t have a cycling airlock and all medbay personal have acces to the upper right room, at wich point windows are the only seperator.

**Comment by kriskog** [comment]
I really want to like this. It has potential, but after playing around with it for a while I came to the conclusion that the current implementation is pretty tedious and not all that fun. Here's my observations and thoughts:
 - Machinery isn't the way to go here. Maybe for some of the stuff like the synthesizers. Building machine frames is tedious, and having to do it for everything got old real quick. Item based similar to atmos pipes, built in the lathe, would feel better.
 - The machinery doesn't have rotation, so I need to push shit around to get the inputs/outputs where I want them. I also need to wrench down the machine frames to insert the board, then unwrench the machine to turn it (clumsily) and wrench it again for it to work.
 - You can't disassemble any of the machinery. Screwdriving just hits them.
 - Outputs/pill press/etc need an on/off setting, to prevent having to dump the first load or having contaminated pills. You also need to be able to purge the pill press.
 - The machinery that look flat (filter, acclimator) should let you walk across it.
 - I couldn't unwrench the reaction chamber, so once I put it down I was stuck with it right there.
 - I never actually did figure out how I could make something as simple as multiver/charcoal. This requires you to mix table salt, then oil, then heat. Trying to mix oil first then table salt will make some phenol and fuck everything up. I also couldnt make the acclimator heat anything. It would make sense to have some simple tanks with a separate output amount setting that I could add along the pipes to make the intermediarys. I have no idea if this is what the reaction chamer is supposed to be, I couldnt get it working. 
 - Plumbing requires A LOT of space. Automatically linking pipes is the main reason for this.

There's probably more, but this is the first impression.

**Comment by optimumtact** [comment]
You can literally layer pipes by layer and colour, if you actually need lots of space you're a brainlet

**Comment by kriskog** [comment]
> 
> 
> You can literally layer pipes by layer and colour

How? And dont you say crayons.
No but really, if thats possible it makes it easier. But actually doing it isn't obvious enough that I intuitively figured it out in the half hour I spent with it. So that should be more clear.

**Comment by Time-Green** [comment]
> I really want to like this. It has potential, but after playing around with it for a while I came to the conclusion that the current implementation is pretty tedious and not all that fun. Here's my observations and thoughts:
> 
> * Machinery isn't the way to go here. Maybe for some of the stuff like the synthesizers. Building machine frames is tedious, and having to do it for everything got old real quick. Item based similar to atmos pipes, built in the lathe, would feel better.
> * The machinery doesn't have rotation, so I need to push shit around to get the inputs/outputs where I want them. I also need to wrench down the machine frames to insert the board, then unwrench the machine to turn it (clumsily) and wrench it again for it to work.
> * You can't disassemble any of the machinery. Screwdriving just hits them.
> * Outputs/pill press/etc need an on/off setting, to prevent having to dump the first load or having contaminated pills. You also need to be able to purge the pill press.
> * The machinery that look flat (filter, acclimator) should let you walk across it.
> * I couldn't unwrench the reaction chamber, so once I put it down I was stuck with it right there.
> * I never actually did figure out how I could make something as simple as multiver/charcoal. This requires you to mix table salt, then oil, then heat. Trying to mix oil first then table salt will make some phenol and fuck everything up. I also couldnt make the acclimator heat anything. It would make sense to have some simple tanks with a separate output amount setting that I could add along the pipes to make the intermediarys. I have no idea if this is what the reaction chamer is supposed to be, I couldnt get it working.
> * Plumbing requires A LOT of space. Automatically linking pipes is the main reason for this.
> 
> There's probably more, but this is the first impression.

I can fix all of that, thank you for your feedback.

**Comment by Time-Green** [comment]
Anyone have an idea for an alternative implementation of the machinery though? I prefer not to start snowflaking, but something like an RPD is too little and protolathe boards is too much.

**Comment by nemvar** [comment]
ugh, r-wall spam

**Comment by CRITAWAKETS** [comment]
Make an RCD equivalent for the machinery. It takes in metal+glass and it plonks down the machine.

**Comment by Time-Green** [comment]
> ugh, r-wall spam

If only we all shared your argueing skills

@CRITAWAKETS people hate RPD’s, but RCD with just plumbing stuff is actually a good idea

**Comment by Mey-Ha-Zah** [comment]
Time, you know construction site's console. Maybe use something similar to that. Its a cool feature that doesn't get used very much.

**Comment by Krysonism** [comment]
Please don't give doctors chem access.

**Comment by MCterra10** [comment]
@Krysonism yeah but it's not chem any more, it's the apothecary

**Comment by Time-Green** [comment]
@CRITAWAKETS I implemented your RCD idea

@kriskog Your feedback was extremely useful. And uhh colors wasnt crayons, I just forgot.
The only thing I didn't change from your feedback was the reagent chamber, since while it's obviously too uninituituve, it does work.  I'll write a full explanation to all the equipment once I'm finished, since alot of it is difficult to bring into short sentences.

It's just map changes and respriting the multi layer duct from here, I'll probably have the whole thing merge/review ready tomorrow.

**Comment by ExcessiveUseOfCobblestone** [comment]
> Krysonism yeah but it's not chem any more, it's the apothecary

can you explain what this means in practical terms? I'm not a fan of just giving the dispenser to MDs if it's functionally the same as it is currently.

**Comment by Time-Green** [comment]
Ready for review

**Comment by Time-Green** [comment]
Ready for review. I'm gonna go write a wiki article in the meantime

**Comment by TheChosenEvilOne** [comment]
Few things I noticed while testing this:
The Plumbing RCD has no visual display like the RCD has for how many matter units it has left.
Can't deconstruct the plumbing machinery with the plumbing RCD, or well with anything on that matter.
There has to be a pipe between machinery or they won't transfer reagents.
Chemical Acclimators do not seem to empty when the acceptable temperature difference is set to 0.
There does not seem to be anyway to make pills with a certain ratio of reagents.
The APC in apothecary seems to be still linked to the chemistry area.

**Comment by Time-Green** [comment]
> Few things I noticed while testing this:
> 
> 1. The Plumbing RCD has no visual display like the RCD has for how many matter units it has left.
> 2. Can't deconstruct the plumbing machinery with the plumbing RCD, or well with anything on that matter.
> 3. There has to be a pipe between machinery or they won't transfer reagents.
> 4. Chemical Acclimators do not seem to empty when the acceptable temperature difference is set to 0.
> 5. There does not seem to be anyway to make pills with a certain ratio of reagents.
> 6. The APC in apothecary seems to be still linked to the chemistry area.
> 7. The pill press should have internal storage for the pills because the pills can quite quickly pile up. (for reference I had a server up for seven minutes, in that time I made a meth factory and now there are 610 pills on a single turf)
> 8. Can't remove filters from the filter and the filters seem to be taking in reagents from the top always.
> 
> But hey, meth.
> picture

1. Good one, I'll add it.

2. You weld the machine to take it apart. I'll see about adding it to the plumbing RCD.

3. Yeah, I was aware of this from the beginning and forgot to document it. I agree it's clumsy and I do plan on getting it fixed eventually.

4. Chemical heaters take ages to actually get there. When it says it's at 473, it is at 473.99999 something, since it slows down constantly to get there precisely. This is unnoticable in the normal chem heater. It's why I added the acceptable temperature difference. I'll try and round it if reacting does so aswell.

5. Intended. You should connect a reaction chamber or filter if you need it to be more precise.

6. fuck

7. I can see why that would not be ideal. I'll make it stop at 10 pills and then switch to storing it untill place is made. 

8. I'll take a look at filters, they're not supposed to do that

**Comment by TheChosenEvilOne** [comment]
Ill test this again tomorrow, but there is a lamp floating in air in chemistry.

**Comment by TheChosenEvilOne** [comment]
Some more testing:
1. The fluid ducts get used when clicking a turf with a fluid duct on it.
2. Putting in and taking out reagents with the gates does the attack animation.
3. When using a beaker on an output gate, it says "The output gate is empty and can't be refilled!"
4. (probably not really an issue but CLF3 formation breaks the reaction chamber)
5. The mentioned lamp floating in air at the grinder

Otherwise seems pretty good

**Comment by Angustmeta** [comment]
I did a few hours of testing as well with the latest commits. It's a fun system to play around with, and rather relaxing compared to old chemistry I'd say. I got a few observations:

1. I couldn’t get two differently colored ducts to work when connected with click dragging+wrench in hand:


2. Got a "script error" when trying to remove "10 chlorine, 10 phosphorus, 10 radium" programming from a reaction chamber, while it was filling. I think it happened when I clicked the last one so probably related to my next point. I also heard the reaction bubble sound the same time:


(Row 7, Character 2352, don't ask)

3. Removing the last requested reagent from a reaction chamber doesn’t refresh UI. You have to close and re-open the menu to see the last “chlorine: 10” gone etc. Possibly local issue, caused by the script error above? 

4. After creating plastic sheets in a reaction chamber with 50 oil, 20 sulphuric acid and 30 heated ash, the reaction chamber gets stuck on “Current status: Emptying” permanently. It also duped 50 non-existing oil into the output gate. Emptying the output gate didn’t make the reaction chamber work again. Rebuilding it made it work but the same thing happens again if it makes new plastic sheets.

**Comment by Time-Green** [comment]
> Some more testing:
> 
> 1. The fluid ducts get used when clicking a turf with a fluid duct on it.
> 2. Putting in and taking out reagents with the gates does the attack animation.
> 3. When using a beaker on an output gate, it says "The output gate is empty and can't be refilled!"
> 4. (probably not really an issue but CLF3 formation breaks the reaction chamber)
> 5. The mentioned lamp floating in air at the grinder
> 
> Otherwise seems pretty good

1. Woops, I never noticed that
2. Looks kinda natural, but yeah I guess it's hindering
3. I'll look into that
4. It's an issue. A chemical heater should ideally not be destroyed by heat
5. It's amazing what the human eye can miss

**Comment by Time-Green** [comment]
> I did a few hours of testing as well with the latest commits. It's a fun system to play around with and rather relaxing compared to old chemistry I'd say. I got a few observations:
> 
> 1. I couldn’t get two differently colored ducts to work when connected with click dragging+wrench in hand:
>    
> 2. Got a "script error" when trying to remove "10 chlorine, 10 phosphorus, 10 radium" programming from a reaction chamber, while it was filling. I think it happened when I clicked the last one so probably related to my next point. I also heard the reaction bubble sound the same time:
> 
> 
> (Row 7, Character 2352, don't ask)
> 
> 1. Removing the last requested reagent from a reaction chamber doesn’t refresh UI. You have to close and re-open the menu to see the last “chlorine: 10” gone etc. Possibly local issue, caused by the script error above?
> 2. After creating plastic sheets in a reaction chamber with 50 oil, 20 sulphuric acid and 30 heated ash, the reaction chamber gets stuck on “Current status: Emptying” permanently. It also duped 50 non-existing oil into the output gate. Emptying the output gate didn’t make the reaction chamber work again. Rebuilding it made it work but the same thing happens again if it makes new plastic sheets.

1. Weird, I'll look into it
2. That's a new one
3. I knew about this one actually. I never managed to fix it and I can't find out why it's doing that
4. I heard someone else talk about that. I'll add a safety net

I'll get on all these reviews probably somewhere in the weekend

**Comment by davethwave** [comment]
This seems like a great PR and can't wait until the first test-merge. Also, for the chem rcd I've heard of. Does it just take metal and glass like normal ones? Plastic could be a fun option as a requirement for this. Also might be nice to be able to build a giant containment tank that takes 3x3 amount of space for long term chem storage or something. Though you would have to worry about it being used as a big chem bomb. Another idea that could be nice is a different machine that bottles chems in a 30u container that takes 200 glass per bottle so it has to be refilled once in a while.

**Comment by FloranOtten** [comment]
I tried it out, and it's glorious. I did encounter some problems, though. While making multiver, my acclimator broke. After it cooked oil into ash, it refused to output or input, simply sitting still with 200u of ash inside. I don't know if it's related to other people's reactions breaking, but I'll put it here anyway.

**Comment by Time-Green** [comment]
> I tried it out, and it's glorious. I did encounter some problems, though. While making multiver, my acclimator broke. After it cooked oil into ash, it refused to output or input, simply sitting still with 200u of ash inside. I don't know if it's related to other people's reactions breaking, but I'll put it here anyway.

Did you properly set the allowed temperature difference variable? It'll only start emptying till it hits the target temp minus allowed temperature difference. Not setting allowed temperature difference will NEVER empty it due to the way heating code works.

**Comment by kriskog** [comment]
Seems like a common mistake, can you add a tooltip/make it more intuitive and/or prevent people from setting too rigid limits?

**Comment by TheChosenEvilOne** [comment]
perhaps you should give it a default value so it does not happen.

**Comment by MarioWizard119** [comment]
As someone who mains both atmos tech and chemist, this PR makes me very, very happy. Plumbing seems to be my favorite two jobs rolled into one awesome meth factory.

This PR will do more to settle the balance feud between Chemist and MD than removing chems or crapmeds ever will.

Surgery terrible? Doctors left with no purpose cause chems are better at healing from every perspective? Brute force the balance feud and give doctors apothecary/old chemistry access! They can make their meds and give them too!

Chemistry system completely brainless? Chemists just want to experiment with things and have things to experiment with? Splice chem with GLORIOUS ATMOSIA and you get this lovely PR right here, and shove them in the back of medical where no one will ever find them and their concussion explosives factory.

Though as much as I love the new system, I do have a few minor suggestions. It would be nice if the floor in between the surgery observation room and the chemistry entrance had white and orange floor tiles to make the new chem lab seem less cut off from the rest of Medbay. Other “maint crossroad” rooms like that have full floor tiles with department colors, like the ones in RnD, cargo, Virology, etc. Also, you may want to add a few more chem dispensers to the apothecary, or else you’ll probably get a similar situation with Xenobiology or the Engineering hardsuits where everyone is fighting for a limited supply of something that greatly expands their options and makes their gameplay more fun. And I know it’s probably already been said, but an option to put chems in bottles or patches (or maybe even syringes or grenades with upgrades) in the pill press would be nice, since with the new meds it’s most advisable to bottle your chems to sip a little at at time instead of taking one huge liver-destroying pill.

Regardless of my complaints, this PR looks amazing and I can’t wait to try it. Godspeed with the bug fixes and best of luck getting it merged. I hope it gets added soon.

**Comment by davethwave** [comment]
Actually, I have a very good idea, how about a kind of chem trash can of sorts to get rid of extra chems that would normally build up.

**Comment by FloranOtten** [comment]
> Actually, I have a very good idea, how about a kind of chem trash can of sorts to get rid of extra chem bi-products that would normally build up.

You mean a chem disposer?

**Comment by Time-Green** [comment]
> Actually, I have a very good idea, how about a kind of chem trash can of sorts to get rid of extra chem bi-products that would normally build up.

Something like a plumbing chemical disposer? That's an amazing idea!

I already added them more than a month ago though

**Comment by Time-Green** [comment]
Plumbing is merge ready again. All bug reports have been fixed or deemed a feature.

**Comment by FloranOtten** [comment]
🚯

**Comment by Time-Green** [comment]
Conflicts fixed. My acclimator changes were moved to tgui-next aswell.

**Comment by optimumtact** [comment]
buckled up cause we about to rock out

**Comment by AnonymousNow** [comment]
Ahh, plumbing is finally happening, eh? Been a fair few years in the making, surely. 

I don't think making the old maintenance bar a dead end is wise, though.

**Comment by FloranOtten** [comment]
> Ahh, plumbing is finally happening, eh? Been a fair few years in the making, surely.
> 
> I don't think making the old maintenance bar a dead end is wise, though.
 
It always was. The entrance is just slightly more secluded.

**Comment by Dorsisdwarf** [comment]
This seems like it would be a really good use of multiZ when TCEO finishes prodding it into a useable state

**Comment by AnonymousNow** [comment]
> 
> 
> This seems like it would be a really good use of multiZ when TCEO finishes prodding it into a useable state

Suddenly, SS13 gets a sewer level.

**Comment by Angustmeta** [comment]
From merged version:
1. I can’t get two differently colored ducts to work when connected with click dragging+wrench in hand (same issue as before). 

2. Removing the last requested reagent from a reaction chamber doesn't work, and trying makes you unable to add any new rules. 

3. Reactions that consume chems (foam etc) will break the reaction chamber (same issue as before).

**Comment by Time-Green** [comment]
> From merged version:
> 
> 1. I can’t get two differently colored ducts to work when connected with click dragging+wrench in hand (same issue as before).
> 2. Removing the last requested reagent from a reaction chamber doesn't work, and trying makes you unable to add any new rules.
> 3. Reaction that made foam broke the reaction chamber (same issue as before, with plastic sheets).

1. The fuck? Did I instantly break it again?

2. It does work, but you need to close and reopen it. I’ll just update the whole UI

3. I never managed to reproduce it. Did you do anything special?

I’m sorry to make you sound like a broken record, I’ll get on it in my spare time. Also please make an issue for future bugs, it’s easier to track like that

**Comment by Angustmeta** [comment]
1. Don't know but I still can't get it to work. 
2. Actually I can't remove any rule at all from the reaction chamber, not even if I close and reopen the menu afterwards. I know I could do that in the test a few days ago. 
3. See this image. 5u water from the left. 5u fluorosurfactant from the right. The reaction chamber in the middle creates foam once and then nothing more happens. 

I tried connecting a disposer to it afterwards but it still didn't start producing foam again:

**Comment by PKPenguin321** [comment]
>Suddenly, SS13 gets a sewer level.

You joke, but an underground maint/sewer z-level would be fucking sweet

**Comment by MarioWizard119** [comment]
> You joke, but an underground maint/sewer z-level would be fucking sweet

Assistant land, an entire realm dedicated to the grey tide

**Pull request merged**
